rfc/rfc-0020-resources-update-after-success.md

```diff
@@ -0,0 +1,91 @@
+# Draft RFC 20 Update Resources Only After Success/Failure
```

Comment:
"Success" and "Failure" are states not typically associated with [level-triggered reconciliation](https://hackernoon.com/level-triggering-and-reconciliation-in-kubernetes-1f17fe30333d). Level reconciliation is fundamental to the nature of Kubernetes where the system will continuously try to align the current state with the desired state. Faults can come and go based on the same desired state.

Reply:
I'd argue that:
1. Conditions are an indication that k8s understands the importance of indicating whether a current state is "Good".
2. We're trying to deal with resources that include an additional fudge to level-triggered reconciliation; the reporting of the most recent "good" output. This includes Runnable's outputs, kpacks `latestImage`. These are based not simply on the spec submitted and the state of the cluster; but also on historical state that is no longer observable.

Reply:
> We're trying to deal with resources that include an additional fudge to level-triggered reconciliation

Every resource that a supply chain is directly templating needs to be level-triggered. Both the Ruannable and the kpack Image resources are level-triggered. Those resource internally manage edge-triggered state, but they themselves are not edge-triggered.

I think what you're struggling with is similar to a Deployments dual conditions of Available and Progressing. While the resource is progressing (rolling out an update), it may or may not also be available (able to receive traffic). While it is progressing, the Pods that are actually receiving traffic may be from the previous desired state or the new desired state.

There aren't a lot of conventions in this space, and they are widely applied which will make it very difficult to have confidence you're looking for.

Reply:
I was discussing this in private but I want to add that, best efforts, a resource might repeatedly go into "failed" state but eventually be "successful". If we stop at "failed" we run the risk of having no way to have the entire supply chain become eventually consistant. We're up against specific implementations of resources, but our goal should be:

* Given a _valid_ `workload` and latest source/image commit
* where _valid_ means the commit should traverse the supply chain (there's nothing wring with it)
* the supply chain is eventually consistent and provides final outputs.

This means the supply chain and the resources controlled by it should be resilient to partition errors and cluster misconfigurations, that when rectified, result in the commit becoming a config and image (or deployed app)

Reply:
That is consistent with the core proposal here, to read only when an object is in a successful state. Given that rule, there is no interference with an object's ability to eventually reconcile successfully.

It _is_ a good argument not to use [this strategy](https://github.com/vmware-tanzu/cartographer/blob/rfc-0020-template-success/rfc/rfc-0020-resources-update-after-success.md#how-to-ensure-the-most-recent-successful) to rollback updates. This is really a separate concern, I'm going to split it into a separate RFC.

Reply:
Yup it was definitely muddying the waters. When you pull it out please add a link to my concern above in the "drawbacks" section. I think it's too easy to lose sight of eventual consistency when we dig into specific motiviations/user-stories.